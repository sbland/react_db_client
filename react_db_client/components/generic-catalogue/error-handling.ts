export class GenericCatalogueError extends Error {
  name: 'GenericCatalogueError';
  error?: Error;
  constructor(msg: string, error?: Error) {
    super(msg);
    this.error = error;
    // Set the prototype explicitly.
    Object.setPrototypeOf(this, GenericCatalogueError.prototype);
  }
}
