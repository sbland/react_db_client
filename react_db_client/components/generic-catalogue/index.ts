export { GenericCatalogue } from './generic-catalogue';
