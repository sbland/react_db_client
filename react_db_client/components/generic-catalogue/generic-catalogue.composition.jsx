import React from 'react';
// import { PopupPanel } from '@samnbuk/react_db_client.components.popup-panel';
import { ItemEditor } from '@samnbuk/react_db_client.components.item-editor';
import { GenericCatalogue } from './generic-catalogue';
import { PopupPanel, PopupPanelConnector } from '@samnbuk/react_db_client.components.popup-panel';

import { demoHeadingsData, demoHeadingsDataSimple, demoResults } from './demo-data';

const ItemEditorPopup = PopupPanelConnector(ItemEditor, 'root', true, 'onCancel', {title: 'Demo Item Editor'});
const PopupPanelConnected = (props) => <PopupPanel popupRoot="root" {...props} />;

const defaultProps = {
  id: 'demo-id',
  itemName: 'Demo Item',
  collection: 'DEMO COLLECTION',
  additionalFilters: [],
  availableFilters: {},
  resultsHeadings: demoHeadingsDataSimple,
  editorHeadings: demoHeadingsData,
  additionalSaveData: {},
  ItemEditor: ItemEditorPopup,
  apiDeleteDocuments: async () => alert('Deleted DOcument'),
  PopupPanel: PopupPanelConnected,
  notificationDispatch: alert,
  asyncGetDocument: async () => Object.values(demoResults)[0],
  asyncGetDocuments: async () => Object.values(demoResults),
  asyncPutDocument: async () => alert('Put doc'),
  asyncPostDocument: async () => alert('Post doc'),
  asyncDeleteDocument: async () => {},
  asyncDeleteDocuments: async () => {},
  componentMap: {},
  onError: () => {},
};

export const BasicGenericCatalogue = () => <GenericCatalogue {...defaultProps} />;

const popupRoot = document.createElement('div', { id: 'popup-root' });
const _PopupItemEditor = (props) => (
  // <PopupPanel isOpen handleClose={() => setShowEditor(false)} popupRoot={popupRoot}>
  <ItemEditor {...props} />
  // </PopupPanel>
);

export const PopupItemEditor = () => (
  <GenericCatalogue {...defaultProps} ItemEditor={_PopupItemEditor} />
);
