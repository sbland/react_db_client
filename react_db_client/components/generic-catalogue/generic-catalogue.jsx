import React, { useState, useEffect, useCallback } from 'react';
import PropTypes from 'prop-types';

import { SearchAndSelect } from '@samnbuk/react_db_client.components.search-and-select';
import { AreYouSureBtn } from '@samnbuk/react_db_client.components.are-you-sure-btn';
import { ItemEditor as _ItemEditor } from '@samnbuk/react_db_client.components.item-editor';
import { Emoji } from '@samnbuk/react_db_client.components.emoji';


export const GenericCatalogue = ({
  itemName,
  collection,
  additionalFilters,
  resultsHeadings,
  editorHeadings,
  additionalSaveData,
  availableFilters,
  ItemEditor,
  errorCallback,
  apiGetDocuments,
  apiDeleteDocument,
  PopupPanel,
}) => {
  const [showEditor, setShowEditor] = useState(false);
  const [selectedUid, setSelectedUid] = useState(null);
  const [reloadDatakey, setReloadDataKey] = useState(0);
  const [handleDelete, setHandleDelete] = useState(false);

  // const { call: copySet } = useAsyncRequest({
  //   args: [],
  //   callFn: apiCopyDocument,
  //   callOnInit: false,
  //   callback: () => {
  //     setReloadDataKey((prev) => prev + 1);
  //     alert('Successfully Copied Set');
  //   },
  // });

  useEffect(() => {
    if (handleDelete) {
      setHandleDelete(false);
      apiDeleteDocument(collection, selectedUid)
        .then(() => {
          setReloadDataKey((prev) => prev + 1);
        })
        .catch(() => {
          errorCallback(`Error deleting ${itemName}`, 'error');
        });
    }
  }, [handleDelete, selectedUid, errorCallback, collection, itemName]);

  // TODO: Implement duplicate
  // const handleDuplicate = (uid) => {
  //   const fromCollection = COLLECTION;
  //   const fromUid = uid;
  //   const toCollection = COLLECTION;
  //   const toUid = generateUid(COLLECTION);
  //   const additionalData = { label: toUid };
  //   copySet([fromCollection, fromUid, toCollection, toUid, additionalData]);
  // };

  const searchFn = useCallback(
    async (filters, sortBy, searchString) => {
      const filtersModified = [...additionalFilters, ...filters];
      const sortByA = sortBy || 'label';
      const docs = await apiGetDocuments(
        collection,
        filtersModified,
        // TODO: Filter preview only headings out
        resultsHeadings.map((item) => item.uid),
        sortByA,
        searchString,
        false,
        'all'
      );
      return docs;
    },
    [collection, additionalFilters, resultsHeadings]
  );

  return (
    <>
      {showEditor && (
        <ItemEditor
          inputUid={selectedUid}
          isNew={!selectedUid && true}
          additionalData={additionalSaveData}
          params={editorHeadings}
          collection={collection}
          onSubmitCallback={(uid) => {
            notificationDispatch(openMessage(`${itemName} saved`));
            setSelectedUid(uid);
            setShowEditor(false);
            setReloadDataKey(reloadDatakey + 1);
          }}
        />
      )}
      <div className="GenericCatalogueFunc_Wrap sectionWrapper">
        <section>
          <button
            type="button"
            className="button-two"
            onClick={() => {
              setSelectedUid(null);
              setShowEditor(true);
            }}
          >
            Create New {itemName}
          </button>
        </section>
        <section style={{ width: '100%' }}>
          <p>
            <b>Adjust the filters below to search for {itemName}s</b>
          </p>
          <SearchAndSelect
            key={reloadDatakey}
            searchFunction={searchFn}
            initialFilters={additionalFilters}
            availableFilters={availableFilters}
            handleSelect={setSelectedUid}
            // TODO: Filter preview only headings
            headings={resultsHeadings}
            previewHeadings={resultsHeadings}
            autoWidth={false}
            autoUpdate
            showRefreshBtn
            showSearchField
            loadOnInit={false}
            noEmptySearch
            allowSelectionPreview
          />
          <div>
            <button
              type="button"
              onClick={() => setShowEditor(true)}
              disabled={selectedUid === null}
            >
              Edit Selected {itemName}
            </button>
            <AreYouSureBtn
              onConfirmed={() => {
                setHandleDelete(true);
              }}
              message={`Delete ${itemName}`}
              disabled={selectedUid === null}
              btnText={<Emoji emoj="🗑️" label="Delete" />}
              PopupPanel={PopupPanel}
            />
          </div>
        </section>
      </div>
    </>
  );
};

GenericCatalogue.propTypes = {
  itemName: PropTypes.string.isRequired,
  collection: PropTypes.string.isRequired,
  additionalFilters: PropTypes.arrayOf(
    PropTypes.shape({
      uid: PropTypes.string.isRequired,
      field: PropTypes.string.isRequired,
      operator: PropTypes.string,
      value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    })
  ),
  resultsHeadings: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
  editorHeadings: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
  additionalSaveData: PropTypes.shape({}),
  availableFilters: PropTypes.objectOf(
    PropTypes.shape({
      uid: PropTypes.string.isRequired,
      label: PropTypes.string.isRequired,
      type: PropTypes.string.isRequired,
      required: PropTypes.bool,
      group: PropTypes.number,
    })
  ).isRequired,
  ItemEditor: PropTypes.elementType,
  apiGetDocuments: PropTypes.func.isRequired,
  apiDeleteDocuments: PropTypes.func.isRequired,
  PopupPanel: PropTypes.elementType,
};

GenericCatalogue.defaultProps = {
  additionalFilters: [],
  additionalSaveData: {},
  ItemEditor: _ItemEditor,
  PopupPanel: () => {},
};
