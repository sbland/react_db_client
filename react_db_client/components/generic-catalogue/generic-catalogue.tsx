import React, { useState, useEffect, useCallback } from 'react';
import PropTypes from 'prop-types';

import { TComponentMap } from '@form-extendable/lib';
import {
  CustomParser,
  IHeading,
  ISearchAndSelectProps,
  SearchAndSelect,
} from '@react_db_client/components.search-and-select';
import { AreYouSureBtn } from '@react_db_client/components.are-you-sure-btn';
import {
  FilterObjectClass,
  FilterOption,
  IDocument,
  TAsyncCopyDocument,
  TAsyncDeleteDocument,
  TAsyncGetDocument,
  TAsyncGetDocuments,
  TAsyncPostDocument,
  TAsyncPutDocument,
  Uid,
} from '@react_db_client/constants.client-types';
import { ItemEditor as ItemEditorDefault } from '@react_db_client/components.item-editor';
import { useAsyncRequest } from '@react_db_client/async-hooks.use-async-request';
import { Emoji } from '@react_db_client/components.emoji';
import { generateUid } from '@react_db_client/helpers.generate-uid';

// Link these to actual types
export interface IItemEditorProps {}
export interface IPopupPanelProps {}

export interface IGenericCatalogueProps<ResultType extends IDocument> {
  id: Uid;
  itemName: string;
  collection: string;
  additionalFilters?: FilterObjectClass[];
  customSort?: (a: ResultType, b: ResultType) => -1 | 0 | 1;
  resultsHeadings: IHeading[];
  editorHeadings: IHeading[];
  additionalSaveData?: object;
  availableFilters: { [key: string]: FilterOption<any, boolean> };
  ItemEditor: typeof ItemEditorDefault;
  errorCallback?: (message: string, e: Error) => void;
  PopupPanel: IPopupPanelProps;
  notificationDispatch: (message: string) => void;
  customParsers?: { [k: string]: CustomParser };
  previewHeadings?: IHeading[];
  asyncGetDocument: TAsyncGetDocument<ResultType>;
  asyncPutDocument: TAsyncPutDocument<ResultType>;
  asyncPostDocument: TAsyncPostDocument<ResultType>;
  asyncGetDocuments: TAsyncGetDocuments<ResultType>;
  asyncDeleteDocument: TAsyncDeleteDocument;
  asyncCopyDocument: TAsyncCopyDocument;
  componentMap: TComponentMap;
  closePopupOnItemSave?: boolean;
  sasProps?: Partial<ISearchAndSelectProps<ResultType>>;
}

/**
 * Generic catalogue wrapper for searching and editing documents from the api
 * @param id
 * @param itemName
 * @param collection
 * @param additionalFilters
 * @param customSort
 * @param resultsHeadings
 * @param editorHeadings
 * @param additionalSaveData
 * @param availableFilters
 * @param ItemEditor
 * @param errorCallback
 * @param PopupPanel
 * @param notificationDispatch
 * @param customParsers
 * @param previewHeadings
 * @param componentMap Map of field component for item editor
 * @param {async function} asyncGetDocument async function to get a document
 * - @argument { string } collection target database collection
 * - @argument { string } uid id of document
 * - @argument { Array[string] } schema fields to request from db
 * - @argument { bool } populate if true populate nested docs
 * @param {async function} asyncGetDocuments async function to search for multiple documents
 * - @argument { string } collection target database collection
 * - @argument { List[FilterObjectClass] } filters list of filters to apply to search
 * - @argument { List[string] } schema list of fields to request
 * - @argument { string } sortBy Field to sort by
 * - @argument { string } textsearch string to use for text(Keyword) searching
 * - @argument { bool } reverseSort if true reverse sorting
 * - @argument { bool } populate if true populate nested docs
 * @param {async function} asyncPutDocument async function to update a document
 * - @argument { string } collection target database collection
 * - @argument { string } uid id of document
 * - @argument { object } data data to upload
 * @param {async function} asyncPostDocument async function to add a new document
 * - @argument { string } collection target database collection
 * - @argument { string } uid id of document
 * - @argument { object } data data to upload
 * @param {async function} asyncDeleteDocument async function to delete a document
 * - @argument { string } collection target database collection
 * - @argument { string } uid id of document
 * @param {async function} asyncCopyDocument async function to copy a document
 * - @argument { string } fromCollection db collection to copy from
 * - @argument { string } fromUid db uid to copy
 * - @argument { string } toCollection db collection to copy to
 * - @argument { string } toUid new document id
 * - @argument { object } additionalData additional data to add to copied doc
 * @returns
 */
export const GenericCatalogue = <ResultType extends IDocument>({
  id,
  itemName,
  collection,
  additionalFilters,
  customSort,
  resultsHeadings,
  editorHeadings,
  additionalSaveData,
  availableFilters,
  ItemEditor,
  errorCallback,
  PopupPanel,
  notificationDispatch,
  customParsers,
  previewHeadings,
  asyncGetDocument,
  asyncGetDocuments,
  asyncPutDocument,
  asyncPostDocument,
  asyncDeleteDocument,
  asyncCopyDocument,
  componentMap,
  closePopupOnItemSave,
  sasProps,
}: IGenericCatalogueProps<ResultType>) => {
  const [showEditor, setShowEditor] = useState(false);
  const [selectedUid, setSelectedUid] = useState<Uid | null>(null);
  const [reloadDatakey, setReloadDataKey] = useState(0);
  const [handleDelete, setHandleDelete] = useState(false);

  /* Handle Delete */
  useEffect(() => {
    if (handleDelete && selectedUid) {
      setHandleDelete(false);
      asyncDeleteDocument(collection, selectedUid)
        .then(() => {
          setReloadDataKey((prev) => prev + 1);
        })
        .catch((e) => {
          if (errorCallback) errorCallback(`Error deleting ${itemName}`, e);
        });
    }
  }, [handleDelete, selectedUid, errorCallback, collection, itemName]);

  const handleSasSelect = useCallback((data?: ResultType | null) => {
    if (data) setSelectedUid(data.uid);
    else setSelectedUid(null);
  }, []);

  const duplicateCallback = useCallback(
    (_, [, , , toUid]) => {
      setReloadDataKey((prev) => prev + 1);
      notificationDispatch(`Successfully Copied ${itemName}`);
      setSelectedUid(toUid);
      setShowEditor(true);
    },
    [notificationDispatch, itemName]
  );

  /* Handle Duplicate */
  const { call: copyItem } = useAsyncRequest<void, any[]>({
    args: [],
    callFn: asyncCopyDocument as any,
    callOnInit: false,
    callback: duplicateCallback,
  });

  const handleDuplicate = (uid) => {
    const fromCollection = collection;
    const fromUid = uid;
    const toCollection = collection;
    const toUid = generateUid(collection);
    const additionalData = { label: toUid };
    copyItem([fromCollection, fromUid, toCollection, toUid, additionalData]);
  };

  /* Handle Search */

  const searchFn = useCallback(
    async (filters, sortBy, searchString) => {
      const sortByA = sortBy || 'label';
      const docs = await asyncGetDocuments(
        collection,
        filters,
        // TODO: Filter preview only headings out
        resultsHeadings.map((item) => item.uid),
        sortByA,
        searchString,
        false,
        'all'
      );
      if (customSort) return docs.sort(customSort);
      return docs;
    },
    [collection, resultsHeadings]
  );

  const onSubmitCallback = useCallback(
    (uid) => {
      notificationDispatch(`${itemName} saved`);
      setSelectedUid(uid);
      if (closePopupOnItemSave) setShowEditor(false);
      setReloadDataKey((prev) => prev + 1);
    },
    [itemName]
  );

  return (
    <>
      {showEditor && (
        <ItemEditor
          inputUid={selectedUid}
          isNew={!selectedUid && true}
          additionalData={additionalSaveData}
          params={editorHeadings}
          collection={collection}
          onSubmitCallback={onSubmitCallback}
          asyncGetDocument={asyncGetDocument}
          asyncPutDocument={asyncPutDocument}
          asyncPostDocument={asyncPostDocument}
          asyncDeleteDocument={asyncDeleteDocument}
          componentMap={componentMap}
          saveErrorCallback={errorCallback}
          // onCancel={() => setShowEditor(false)}
          id={`item_editor_${id}`}
        />
      )}
      <div className={`genericCatalogueFunc_Wrap sectionWrapper genericCatalogue_${id}`}>
        <section>
          <button
            type="button"
            className="button-two genericCatalogue_createNewBtn"
            onClick={() => {
              setSelectedUid(null);
              setShowEditor(true);
            }}
          >
            Create New {itemName}
          </button>
        </section>
        <section style={{ width: '100%' }}>
          <p>
            <b>Adjust the filters below to search for {itemName}s</b>
          </p>
          <SearchAndSelect
            key={reloadDatakey}
            searchFunction={searchFn}
            initialFilters={additionalFilters}
            availableFilters={availableFilters}
            handleSelect={handleSasSelect}
            headings={resultsHeadings}
            previewHeadings={previewHeadings || resultsHeadings}
            customParsers={customParsers}
            autoUpdate
            showRefreshBtn
            showSearchField
            loadOnInit={false}
            noEmptySearch
            allowSelectionPreview
            {...sasProps}
          />
          <div>
            <button
              type="button"
              className="button-two genericCatalogue_editSelectedBtn"
              onClick={() => setShowEditor(true)}
              disabled={selectedUid === null}
            >
              Edit Selected {itemName}
            </button>
            <button
              type="button"
              className="button-one genericCatalogue_duplicateSelectedBtn"
              onClick={() => handleDuplicate(selectedUid)}
              disabled={selectedUid === null}
            >
              Duplicate Selected {itemName}
            </button>
            <AreYouSureBtn
              onConfirmed={() => {
                setHandleDelete(true);
              }}
              message={`Delete ${itemName}`}
              disabled={selectedUid === null}
              btnText={<Emoji emoj="🗑️" label="Delete" />}
              PopupPanel={PopupPanel}
            />
          </div>
        </section>
      </div>
    </>
  );
};

GenericCatalogue.propTypes = {
  id: PropTypes.string.isRequired,
  itemName: PropTypes.string.isRequired,
  collection: PropTypes.string.isRequired,
  additionalFilters: PropTypes.arrayOf(
    PropTypes.shape({
      uid: PropTypes.string.isRequired,
      field: PropTypes.string.isRequired,
      operator: PropTypes.string,
      value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    })
  ),
  customSort: PropTypes.func,
  customParsers: PropTypes.objectOf(PropTypes.func),
  resultsHeadings: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
  editorHeadings: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
  previewHeadings: PropTypes.arrayOf(PropTypes.shape({})),
  additionalSaveData: PropTypes.shape({}),
  availableFilters: PropTypes.objectOf(
    PropTypes.shape({
      uid: PropTypes.string.isRequired,
      label: PropTypes.string.isRequired,
      type: PropTypes.string.isRequired,
      required: PropTypes.bool,
      group: PropTypes.number,
    })
  ).isRequired,
  ItemEditor: PropTypes.elementType,
  PopupPanel: PropTypes.elementType,
  notificationDispatch: PropTypes.func,
  asyncGetDocuments: PropTypes.func.isRequired,
  asyncDeleteDocuments: PropTypes.func.isRequired,
  asyncGetDocument: PropTypes.func.isRequired,
  asyncPutDocument: PropTypes.func.isRequired,
  asyncPostDocument: PropTypes.func.isRequired,
  asyncDeleteDocument: PropTypes.func.isRequired,
  componentMap: PropTypes.objectOf(PropTypes.elementType),
  errorCallback: PropTypes.func,
  closePopupOnItemSave: PropTypes.bool,
};

GenericCatalogue.defaultProps = {
  additionalFilters: [],
  additionalSaveData: {},
  customSort: null,
  previewHeadings: [],
  customParsers: {},
  ItemEditor: ItemEditorDefault,
  PopupPanel: () => {},
  notificationDispatch: alert,
  componentMap: {},
  errorCallback: () => {},
  closePopupOnItemSave: false,
};
