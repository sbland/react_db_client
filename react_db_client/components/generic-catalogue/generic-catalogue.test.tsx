import '@samnbuk/react_db_client.testing.enzyme-setup';
import React from 'react';
import { mount } from 'enzyme';
import { render, fireEvent, within, act } from '@testing-library/react';
import { BasicGenericCatalogue } from './generic-catalogue.composition';


const defaultProps = {};

describe('CertificateCatalogue', () => {
  it('Renders', () => {
    mount(<BasicGenericCatalogue />);
  });
  it('Matches Snapshot', () => {
    const component = mount(<BasicGenericCatalogue />);
    const tree = component.debug();
    expect(tree).toMatchSnapshot();
  });
  describe('Functional tests', () => {
    test('should render list of results', async () => {
      const component = render(<BasicGenericCatalogue />);
      const searchInput = component.getByLabelText('search');
      await act(async () => {
        fireEvent.change(searchInput, { target: { value: 'demo' } });
        await new Promise((resolve) => setTimeout(resolve));
      });
      const resultsList = component.getByRole('list');
      const { getAllByRole } = within(resultsList);
      const items = getAllByRole('listitem');
      expect(items.length).toBeGreaterThan(0);
      expect(items.length).toBe(5);
    });
  });
});
