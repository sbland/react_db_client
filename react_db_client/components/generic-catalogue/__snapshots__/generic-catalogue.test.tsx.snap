// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CertificateCatalogue Matches Snapshot 1`] = `
"<BasicGenericCatalogue>
  <GenericCatalogue id=\\"demo-id\\" itemName=\\"Demo Item\\" collection=\\"DEMO COLLECTION\\" additionalFilters={{...}} availableFilters={{...}} resultsHeadings={{...}} previewHeadings={{...}} editorHeadings={{...}} additionalSaveData={{...}} ItemEditor={[Function (anonymous)]} PopupPanel={[Function: PopupPanelConnected]} notificationDispatch={[Function (anonymous)]} asyncGetDocument={[Function: asyncGetDocument]} asyncGetDocuments={[Function: asyncGetDocuments]} asyncPutDocument={[Function: asyncPutDocument]} asyncPostDocument={[Function: asyncPostDocument]} asyncDeleteDocument={[Function: asyncDeleteDocument]} asyncCopyDocument={[Function: asyncCopyDocument]} componentMap={{...}} customSort={{...}} customParsers={{...}} errorCallback={[Function: errorCallback]} closePopupOnItemSave={false}>
    <div className=\\"genericCatalogueFunc_Wrap sectionWrapper genericCatalogue_demo-id\\">
      <section>
        <button type=\\"button\\" className=\\"button-two genericCatalogue_createNewBtn\\" onClick={[Function: onClick]}>
          Create New 
          Demo Item
        </button>
      </section>
      <section style={{...}}>
        <p>
          <b>
            Adjust the filters below to search for 
            Demo Item
            s
          </b>
        </p>
        <SearchAndSelect searchFunction={[Function (anonymous)]} initialFilters={{...}} availableFilters={{...}} handleSelect={[Function (anonymous)]} headings={{...}} previewHeadings={{...}} customParsers={{...}} autoUpdate={true} showRefreshBtn={true} showSearchField={true} loadOnInit={false} noEmptySearch={true} allowSelectionPreview={true} selectionOverride={{...}} allowFilters={true} allowMultiple={false} returnFieldOnSelect=\\"uid\\" searchFieldTargetField={{...}} acceptSelectionBtnText=\\"Accept Selection\\" limitResultHeight={{...}} sortBy=\\"uid\\" reverseSort={false} reloadKey={0} liveUpdate={false} autoWidth={true} labelField=\\"label\\" autoPreview={false}>
          <styled.div>
            <div className=\\"sc-ehvNnt jtUYlD\\">
              <div className=\\"searchAndSelect sas_wrap sectionWrapper\\">
                <section className=\\"sas_filtersSection\\" style={{...}}>
                  <button type=\\"button\\" className=\\"button-one\\" onClick={[Function: onClick]}>
                    Show Preview
                  </button>
                  <FilterPanel filterData={{...}} addFilter={[Function: addFilter]} deleteFilter={[Function: deleteFilter]} updateFilter={[Function: updateFilter]} clearFilters={[Function: handleClearFilters]} updateFieldTarget={[Function: updateFieldTarget]} updateOperator={[Function: updateOperator]} fieldsData={{...}} showPanelOverride={false} floating={false} autoOpenPanel={false} customFilters={{...}} customFiltersComponents={{...}}>
                    <div className=\\"filterManager\\">
                      <button type=\\"button\\" className=\\"button-one openFiltersButton\\" onClick={[Function: onClick]}>
                        Filters
                      </button>
                      <div className=\\"filterPanel_panel \\" style={{...}}>
                        <FiltersList filterData={{...}} deleteFilter={[Function: deleteFilter]} updateFilter={[Function: updateFilter]} fieldsData={{...}} customFilters={{...}} customFiltersComponents={{...}} updateFieldTarget={[Function: updateFieldTarget]} updateOperator={[Function: updateOperator]}>
                          <div>
                            <styled.div>
                              <div className=\\"sc-AHaJN dKqwOJ\\">
                                <Styled(styled.div)>
                                  <div className=\\"sc-khsqcC sc-hTBuwn gTmAxE ixcWmP\\" />
                                </Styled(styled.div)>
                                <Styled(styled.div)>
                                  <div className=\\"sc-khsqcC sc-jfTVlA gTmAxE feHWhD\\">
                                    <label>
                                      Field
                                    </label>
                                  </div>
                                </Styled(styled.div)>
                                <Styled(styled.div)>
                                  <div className=\\"sc-khsqcC sc-gScZFl gTmAxE jLsgAI\\">
                                    <label>
                                      Operator
                                    </label>
                                  </div>
                                </Styled(styled.div)>
                                <Styled(styled.div)>
                                  <div className=\\"sc-khsqcC sc-lbVpMG gTmAxE bhweQq\\">
                                    <label>
                                      Value
                                    </label>
                                  </div>
                                </Styled(styled.div)>
                              </div>
                            </styled.div>
                            <styled.ul className=\\"filterPanel_filterList\\">
                              <ul className=\\"sc-bCfvAP clBlND filterPanel_filterList\\">
                                No filters
                              </ul>
                            </styled.ul>
                          </div>
                        </FiltersList>
                        <button type=\\"button\\" className=\\"button-one\\" onClick={[Function: onClick]}>
                          Clear Filters
                        </button>
                        <AddFilterButton fieldsData={{...}} returnNewFilter={[Function: returnNewFilter]} customFilters={{...}}>
                          <button type=\\"button\\" className=\\"button-two addFilterBtn\\" onClick={[Function: handleNewFilter]}>
                            Add Filter
                          </button>
                        </AddFilterButton>
                      </div>
                    </div>
                  </FilterPanel>
                  <div style={{...}}>
                    <label>
                      Search: 
                    </label>
                    <input className=\\"searchField\\" style={{...}} aria-label=\\"search\\" type=\\"text\\" placeholder=\\"search...\\" value=\\"\\" onChange={[Function: handleSearchFieldInput]} />
                  </div>
                  <button type=\\"button\\" className=\\"button-reset refreshBtn\\" onClick={[Function: onClick]} style={{...}}>
                    <Emoji emoj=\\"🔄\\" label=\\"refresh\\" className=\\"\\">
                      <styled.span className=\\"emoji \\" role=\\"img\\" aria-label=\\"refresh\\">
                        <span className=\\"sc-jOiSOi kAGlAj emoji \\" role=\\"img\\" aria-label=\\"refresh\\">
                          🔄
                        </span>
                      </styled.span>
                    </Emoji>
                  </button>
                </section>
                <section className=\\"sas_resultsSection\\" style={{...}}>
                  <div className=\\"sas_resultsList\\">
                    <div style={{...}}>
                      <StyledSelectList listInput={{...}} headings={{...}} handleSelect={[Function (anonymous)]} currentSelection={{...}} limitHeight={{...}} selectionField=\\"uid\\" autoWidth={true} customParsers={{...}} liveColumnWidthDragging={false}>
                        <styled.div style={{...}} data-testid=\\"styledSelectList\\">
                          <div style={{...}} data-testid=\\"styledSelectList\\" className=\\"sc-cjibBx hHeELA\\">
                            <styled.div style={{...}}>
                              <div style={{...}} className=\\"sc-jcMfQk dntdJq\\">
                                <styled.div style={{...}}>
                                  <div style={{...}} className=\\"sc-eeMvmM llszaG\\">
                                    UID
                                  </div>
                                </styled.div>
                                <styled.div style={{...}}>
                                  <div style={{...}} className=\\"sc-eeMvmM llszaG\\">
                                    Name
                                  </div>
                                </styled.div>
                                <styled.div style={{...}}>
                                  <div style={{...}} className=\\"sc-eeMvmM llszaG\\">
                                    Count
                                  </div>
                                </styled.div>
                                <styled.div style={{...}}>
                                  <div style={{...}} className=\\"sc-eeMvmM llszaG\\">
                                    Eval
                                  </div>
                                </styled.div>
                              </div>
                            </styled.div>
                            <styled.ul limitHeight={false} style={{...}}>
                              <ul style={{...}} className=\\"sc-cabOPr iZLcJB\\" />
                            </styled.ul>
                            <ColumnWidthManagerRender resizeColumn={[Function: resizeColumn]} resizingColumn={false} liveColumnWidths={{...}} columnEdgePositions={{...}} onMouseDownResizeHandle={[Function: onMouseDownResizeHandle]} mouseOverEdge={[Function: mouseOverEdge]} endDragging={[Function: endDragging]} handlePosition={-1} tableWidth={403}>
                              <styled.div className=\\"columnWidthManager_styles\\" style={{...}}>
                                <div className=\\"sc-cCjUiG dzXSHP columnWidthManager_styles\\" style={{...}}>
                                  <div className=\\"columnWidthManager\\" style={{...}}>
                                    <div style={{...}} onMouseDown={[Function: onMouseDownResizeHandle]} onMouseEnter={[Function: onMouseEnter]}>
                                      <div style={{...}} />
                                    </div>
                                    <div style={{...}} onMouseDown={[Function: onMouseDownResizeHandle]} onMouseEnter={[Function: onMouseEnter]}>
                                      <div style={{...}} />
                                    </div>
                                    <div style={{...}} onMouseDown={[Function: onMouseDownResizeHandle]} onMouseEnter={[Function: onMouseEnter]}>
                                      <div style={{...}} />
                                    </div>
                                    <div style={{...}} onMouseDown={[Function: onMouseDownResizeHandle]} onMouseEnter={[Function: onMouseEnter]}>
                                      <div style={{...}} />
                                    </div>
                                    <div className=\\"columnResizeCanvas\\" style={{...}} onMouseUp={[Function: onMouseUp]} onMouseLeave={[Function: onMouseLeave]} onMouseMove={[Function: onMouseMove]} />
                                    <div className=\\"columnResizeHandle\\" style={{...}} onMouseDown={[Function: onMouseDownResizeHandle]} onMouseUp={[Function: endDragging]} />
                                  </div>
                                </div>
                              </styled.div>
                            </ColumnWidthManagerRender>
                          </div>
                        </styled.div>
                      </StyledSelectList>
                    </div>
                    <div className=\\"sas_resultsList-empty\\">
                      No results found. Try adjusting the filters above.
                    </div>
                  </div>
                </section>
              </div>
            </div>
          </styled.div>
        </SearchAndSelect>
        <div>
          <button type=\\"button\\" className=\\"button-two genericCatalogue_editSelectedBtn\\" onClick={[Function: onClick]} disabled={true}>
            Edit Selected 
            Demo Item
          </button>
          <button type=\\"button\\" className=\\"button-one genericCatalogue_duplicateSelectedBtn\\" onClick={[Function: onClick]} disabled={true}>
            Duplicate Selected 
            Demo Item
          </button>
          <AreYouSureBtn onConfirmed={[Function: onConfirmed]} message=\\"Delete Demo Item\\" disabled={true} btnText={{...}} PopupPanel={[Function: PopupPanelConnected]} className=\\"button-one\\" notes=\\"\\" confirmMessage=\\"Confirm\\">
            <PopupPanelConnected id=\\"areYouSureBtn_popupPanel\\" isOpen={false} handleClose={[Function: handleClose]}>
              <PopupPanel popupRoot=\\"root\\" id=\\"areYouSureBtn_popupPanel\\" isOpen={false} handleClose={[Function: handleClose]} renderWhenClosed={false} />
            </PopupPanelConnected>
            <button disabled={true} onClick={[Function: onClick]} type=\\"button\\" className=\\"areYouSureBtn button-one\\">
              <Emoji emoj=\\"🗑️\\" label=\\"Delete\\" className=\\"\\">
                <styled.span className=\\"emoji \\" role=\\"img\\" aria-label=\\"Delete\\">
                  <span className=\\"sc-jOiSOi kAGlAj emoji \\" role=\\"img\\" aria-label=\\"Delete\\">
                    🗑️
                  </span>
                </styled.span>
              </Emoji>
            </button>
          </AreYouSureBtn>
        </div>
      </section>
    </div>
  </GenericCatalogue>
</BasicGenericCatalogue>"
`;
